sway-lib-std/src/vec.sw

```diff
@@ -0,0 +1,147 @@
+library vec;
+
+use ::alloc::{alloc, realloc};
+use ::intrinsics::size_of;
+use ::context::registers::heap_ptr;
+use ::mem::{read, write};
+use ::option::Option;
+
+struct RawVec<T> {
+    ptr: u64,
+    cap: u64,
+}
```

Comment:
`RawVec<T>` looks a lot like a simple buffer besides growing based on `size_of::<T>()`. We could turn this into `Buffer` and move the growth logic up to `Vec<T>`.

Then the task on https://github.com/FuelLabs/fuels-rs/issues/353 would become adding support for `Vec<T>` and `Buffer` instead of `Vec<T>` and `RawVec<T>`.

`RawVec<T>` can not give us "an array of bytes" since our `u8`s are padded to 8 bytes. (`Vec<u8> == Vec<u64> != Buffer`). `Buffer` would work on raw bytes and unlock more possibilities, one being on the Multicall script where I have my own implementation of `Buffer`: https://gist.github.com/AlicanC/7a02df9e987a1a0007f28ca9730cfd7a#file-buf-rs-L101-L104 (definitely not final code)

cc @adlerjohn @digorithm

Reply:
This is based on how Rust [implements `Vec`](https://github.com/rust-lang/rust/blob/6ec3993ef4a4eb72bc20477fe9a4d92acd53f2c6/library/alloc/src/vec/mod.rs#L401) using [`RawVec`](https://github.com/rust-lang/rust/blob/6ec3993ef4a4eb72bc20477fe9a4d92acd53f2c6/library/alloc/src/raw_vec.rs#L52-L54). A `RawVec` _is_ a buffer, and can be used by collections other than `Vec`, such as a buffer that deals with word-aligned padding nicer.

We should imitate the Rust stdlib as much as possible to make implementation to future things easier.